src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSProxyContext.cs

```diff
@@ -0,0 +1,609 @@
+#endif
+
+        #region Current operation context
+
+#if !FEATURE_WASM_THREADS
+        public static readonly JSProxyContext MainThreadContext = new();
+        public static JSProxyContext CurrentThreadContext => MainThreadContext;
+        public static JSProxyContext CurrentOperationContext => MainThreadContext;
+        public static JSProxyContext PushOperationWithCurrentThreadContext()
+        {
+            // in single threaded build we don't have to keep stack of operations and the context/thread is always the same
+            return MainThreadContext;
+        }
+#else
+
+        // Context of the main thread
+        private static JSProxyContext? _MainThreadContext;
+        public static JSProxyContext MainThreadContext
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _MainThreadContext!;
+            set => _MainThreadContext = value;
+        }
+
+        private sealed class PendingOperation
+        {
+            public JSProxyContext? CapturedContext;
+            public bool Called;
+            public bool Multiple;
+        }
+
+        [ThreadStatic]
+        private static List<PendingOperation>? _OperationStack;
+        private static List<PendingOperation> OperationStack
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                return _OperationStack ??= new();
+            }
+        }
+
+        public static void PushOperationUnknownContext()
+        {
+            var stack = OperationStack;
+            // for SchedulePopOperation()
+            if (stack.Count > 0 && stack[stack.Count - 1].Called)
+            {
+                PopOperation();
+            }
+            // because this is called multiple times for each JSImport
+            if (stack.Count > 0 && stack[stack.Count - 1].Multiple)
+            {
+                return;
+            }
+
+            stack.Add(new PendingOperation() { Multiple = true });
+        }
+
+        public static void PushOperationWithContext(JSProxyContext knownContext)
+        {
+            var stack = OperationStack;
+            // for SchedulePopOperation()
+            if (stack.Count > 0 && stack[stack.Count - 1].Called)
+            {
+                PopOperation();
+            }
+            // because this is called multiple times for each JSImport
+            if (stack.Count > 0 && stack[stack.Count - 1].Multiple)
+            {
+                return;
+            }
+
+            stack.Add(new PendingOperation() { CapturedContext = knownContext, Multiple = true });
+        }
+
+        public static JSProxyContext PushOperationWithCurrentThreadContext()
+        {
+            var stack = OperationStack;
+            // for SchedulePopOperation()
+            if (stack.Count > 0 && stack[stack.Count - 1].Called)
+            {
+                PopOperation();
+            }
+
+            var current = AssertIsInteropThread();
```

Comment:
Shouldn't there be a `.Multiple` check here? I am surprised there isn't one

Reply:
I changed all the places from `PushOperationWithCurrentThreadContext` and `PushOperationWithContext` are called, so to use `InitializeImpl` instaed of `Initialize`. Therefore they don't need `SkipAdditionalPushes` flag.